lib/ide/ast/impl/src/known.rs

```diff
@@ -0,0 +1,130 @@
+//! This module provides KnownAst<T> wrapper over Ast that allows expressing that we already
+//! know what `Shape` variant is being stored within this `Ast` node.
+
+use crate::prelude::*;
+
+use crate::Ast;
+use crate::Shape;
+
+
+
+// =================
+// === Known AST ===
+// =================
+
+/// Wrapper for an AST node of known shape type that we can access.
+/// Use `TryFrom<&Ast>` to obtain values.
+///
+/// Provides `Deref` implementation that allows accessing underlying shape `T` value.
+#[derive(Derivative)]
+#[derivative(Clone(bound=""))]
+#[derive(Debug)]
+pub struct KnownAst<T>(Ast, PhantomData<T>);
```

Comment:
IMO We're getting a little messy here:
1. We have Shape enum for specific types AST nodes
2. Then we have `Ast` structure being polymorphic _any shape_ + some data
3. Then we create KnownAst which restores the specific types.

How I would like to write it (which IMO is more rusty, but less Haskellish):
1. specific Shape enumerations.
2. Ast<T> which adds id and span fields to T. 
3. AstNode trait with id and span getters.
4. AnyAst which is like Box<dyn AstNode>
I'm not really convinced to names, but this is just an example how to make types structure more clear.
I'm also not sure, if we really want to refactor this at our timeline - but I really need to share my thoughts about this code